File: 1020-number-of-enclaves/1020-number-of-enclaves.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        def dfs(i,j,grid):
            if(i<0 or j<0 or i>=len(grid) or j>=len(grid[0]) or grid[i][j]==0):
                return 0
            grid[i][j]=0
            return 1+dfs(i+1,j,grid)+dfs(i-1,j,grid)+dfs(i,j-1,grid)+dfs(i,j+1,grid)
        res=0
        res1=0
        for i in grid:
            res+=sum(i)
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if((i==0 or j==0 or i==len(grid)-1 or j==len(grid[0])-1) and grid[i][j]==1):
                    res1+=dfs(i,j,grid)
        return res-res1

File: 1020-number-of-enclaves/1020-number-of-enclaves.cpp
class Solution {
public:
    int numEnclaves(vector<vector<int>>& grid) {
        int res=0;
        int res1=0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++)
                    res1+=grid[i][j];
        }
        int close_one = 0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if((i==0 || j==0 || i==grid.size()-1 || j==grid[0].size()-1) && grid[i][j]==1)
                    res+= dfs(i,j,grid);
            }
        }
        return res1 - res;
    }
                   
    int dfs(int i,int j,vector<vector<int>> &grid){
        if( i<0 || j<0 || i>=grid.size() || j>=grid[0].size() || grid[i][j]==0)
            return 0;
        
        grid[i][j] = 0;
        
        return 1+dfs(i-1,j,grid)+dfs(i+1,j,grid)+dfs(i,j-1,grid)+dfs(i,j+1,grid);
    }
};